zerver/actions/message_flags.py

```diff
@@ -43,13 +44,19 @@ def do_mark_all_as_read(user_profile: UserProfile) -> int:
     )
     do_clear_mobile_push_notifications_for_ids([user_profile.id], all_push_message_ids)
 
-    msgs = UserMessage.objects.filter(user_profile=user_profile).extra(
-        where=[UserMessage.where_unread()],
+    msgs = (
+        UserMessage.objects.select_for_update(no_key=True)
+        .filter(user_profile=user_profile)
+        .extra(
+            where=[UserMessage.where_unread()],
+        )
+        .order_by("id")
     )
 
-    count = msgs.update(
-        flags=F("flags").bitor(UserMessage.flags.read),
-    )
+    with transaction.atomic():
+        count = msgs.update(
+            flags=F("flags").bitor(UserMessage.flags.read),
+        )
```

Comment:
What does the `transaction.atomic` here do? This line is just a single query.

Reply:
Ah I mistakenly thought any db operation after SELECT FOR UPDATE needs to be wrapped inside `transaction.atomic()` — tested that it's fine without `transaction.atomic()`  — will remove it

Reply:
@timabbott  Actually I discovered that when one uses `select_for_update`, any subsequent operations that evaluate the queryset (ie running the query in the db) will need to be enclosed within `transaction.atomic`

Reply:
@timabbott  Actually I discovered that when one uses `select_for_update`, any subsequent operations that evaluate the queryset (ie running the query in the db) will need to be enclosed within `transaction.atomic`, else Python will throw an error like the following:

```
django.db.transaction.TransactionManagementError: select_for_update cannot be used outside of a transaction.
```

Reply:
OK, thanks for confirming.